### PR TITLE
Rewrite table block to use a simpler RichText value

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Split out from #7890 (RichText Structure). The table block uses the `RichText` component for the whole table (many nested elements) in ordere to take advantage of some TinyMCE table controls. This is not really needed: we can create our own table structure with smaller `RichText` values, create our own table controls based on that, and drop the TinyMCE table plugin. The `RichText` component is also not really made for such huge fields, only one line of text with formatting, or at most multiple lines with formatting (array of block-like elements).

In addition to this, the empty state for the table block shows now some inputs to easily create a table of desired dimensions.

The new format also makes it easier to add more controls later, such as adding a head and foot, and maybe move the controls within table to insert and delete rows and columns. Before we didn't really have that freedom.

## How has this been tested?
Ensure the table block still works as before. All controls should be present and work as expected.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: core-blocks/table/table-block.js
Comment (iseulde) [reply core-blocks/table/table-block.js]: Oh, ignore me, we're still using it in the list block.

**Comment by jasmussen** [comment]
I'm seeing these two things:

screen shot 2018-08-09 at 13 50 09

and

screen shot 2018-08-09 at 13 50 42

In master I'm not seeing the initial prompt to choose how many table cells to insert, nor do I see as many inline table commands:

screen shot 2018-08-09 at 13 51 23

That initial prompt, I think, can work well! I think we'll want to style it a little bit, I can probably help there. The chief change being that we'll want to use $default-font and $default-font-size. 

The extra buttons on the toolbar are fine on the desktop, but how do we make them scale to mobile? I have some ideas for a smart toolbar that pops off items into an overflow menu as space becomes a premium, but in absence of that, perhaps the dropdown was better?

Also, I noticed that this control broke:

screen shot 2018-08-09 at 13 53 54

It's broken in `master` too, but mentioning regardless. The point of this is to make sure columns are the same width, no matter the contents inside.

**Comment by jasmussen** [comment]
Regression with fixed width noted in https://github.com/WordPress/gutenberg/pull/7899/comment#issuecomment-411732824.

**Comment by iseulde** [comment]
> That initial prompt, I think, can work well! I think we'll want to style it a little bit, I can probably help there.

Yes, please! :)

Before an earlier rebase this looked better, it would probably be good if controls added inside blocks somehow magically look good (for other block authors).

> nor do I see as many inline table commands

Sorry I forgot they were inside a dropdown. No problem to change.

**Comment by iseulde** [comment]
Also as mentioned in the summary I added the prompt to quickly be able to create the right table size needed. This is inspired by the TinyMCE editor table button. Otherwise you'd have to start off with 4x4 and insert many rows and columns. It would also be cool to be able to add more as you type of course. If the prompt is not so good we can take it away, it was just to try out something different.

**Comment by jasmussen** [comment]
I dig the prompt, I think it's a perfect use of the block concept. 

I also agree it would be nice if block _UI_ were born with the right font and size, any good idea how to do that?

I don't necessarily prefer the dropdown, but I think it's probably necessary until such a time as we have a smarter toolbar.

**Comment by iseulde** [comment]
> I also agree it would be nice if block UI were born with the right font and size, any good idea how to do that?

I'm not sure, but if it requires different styles for the inspector and inside blocks, maybe it requires different components? But it would be good if the same styles work for both.

**Comment by noisysocks** [comment]
There's an error when compiling assets:

```
[0] ERROR in ./packages/block-library/src/table/edit.js
[0] Module not found: Error: Can't resolve './state' in '/Users/robert/projects/gutenberg/packages/block-library/src/table'
[0]  @ ./packages/block-library/src/table/edit.js 21:0-107 75:20-31 101:22-39 121:22-31 142:22-31 164:22-34 185:22-34
[0]  @ ./packages/block-library/src/table/index.js
[0]  @ ./block-library/index.js
```

**Comment by noisysocks** [comment]
First impressions without having looked at the code:

- Centre-aligning the table doesn't position the table in the centre of the editor. It also makes the text in each cell centred, which is inconsistent with what right-aligning the table does.

table-1

- It would be really nice if pressing <kbd>↑</kbd> and <kbd>↓</kbd> moved focus into the cell that's above or below the currently selected cell. Right now it moves focus into the cell that's left or right of the currently selected cell.

- It might be nice to set a minimum width on each cell so that the placeholder showing and hiding doesn't cause columns to resize. Or maybe just ditch the placeholder.

table-2

- _Add Row Before_ and _Add Column Before_ seem to always insert a row or column at the start of the table instead of above or to the left of the currently selected row or column.

table-3

File: packages/block-library/src/table/edit.js
Comment (noisysocks) [inline packages/block-library/src/table/edit.js]: A 10x10 table means we will have 100 instances of TinyMCE loaded on the page. Will these sorts of uses-cases use too much memory?

File: packages/block-library/src/table/edit.js
Comment (noisysocks) [inline packages/block-library/src/table/edit.js]: I think it would feel nice to auto-select this field when the table block is inserted.

File: core-blocks/table/state.js
Comment (noisysocks) [inline core-blocks/table/state.js]: Splitting these functions out is a great idea 🌈

File: packages/blocks/src/api/parser.js
Comment (noisysocks) [inline packages/blocks/src/api/parser.js]: Can we add unit tests for this new attribute source?

File: packages/block-library/src/table/index.js
Comment (noisysocks) [inline packages/block-library/src/table/index.js]: Will we need to add anything to the `deprecated` attribute or does this generate the same markup as before?

**Review by noisysocks — COMMENTED** [review]
This is some great looking code!

I definitely think this refactor makes the table block easier to work with—it should help us a lot with tackling https://github.com/WordPress/gutenberg/issues/6923.

**Review by talldan — DISMISSED** [review]
This looks good. As mentioned, these changes seem like it would make the proposed additions to the table block much easier. 👍 

As @noisysocks pointed out, I did wonder what the performance implications were in having so many RichText elements. I don't know much about it, but have seen it discussed that we're looking to improve the performance.

I've added a few comments, mostly stylistic things and mostly focusing on the edit file.

File: packages/block-library/src/table/edit.js
Comment (iseulde) [reply packages/block-library/src/table/edit.js]: Yes, sure, we should also be looking into the general performance of this. A table with 100 cells is the same as 100 paragraphs.

File: packages/block-library/src/table/index.js
Comment (iseulde) [reply packages/block-library/src/table/index.js]: In there that should still work. I'll test.

File: core-blocks/table/state.js
Comment (iseulde) [reply core-blocks/table/state.js]: Okay. I don't really care what it's called. It's probably closest to the `reducer.js` files we have.

File: packages/block-library/src/table/edit.js
Comment (iseulde) [reply packages/block-library/src/table/edit.js]: Just for my own information, what's the benefit of naming it?

File: packages/block-library/src/table/edit.js
Comment (iseulde) [reply packages/block-library/src/table/edit.js]: You can have `th` in `tbody` too.

File: packages/block-library/src/table/edit.js
Comment (iseulde) [reply packages/block-library/src/table/edit.js]: Interesting!

**Comment by youknowriad** [comment]
I moved it to `3.7` because it seems there's still some work to do here before shipping. Please bring back to 3.6 if it can make the cut.

**Comment by iseulde** [comment]
> Centre-aligning the table doesn't position the table in the centre of the editor. It also makes the text in each cell centred, which is inconsistent with what right-aligning the table does.

Same in master. Let's not fix all issues here.

> It might be nice to set a minimum width on each cell so that the placeholder showing and hiding doesn't cause columns to resize. Or maybe just ditch the placeholder.

I'll ditch it so it's the same as master.

> It would be really nice if pressing ↑ and ↓ moved focus into the cell that's above or below the currently selected cell. Right now it moves focus into the cell that's left or right of the currently selected cell.

Yes, that's a limitation atm, maybe worth fixing writing flow in a separate PR. It's not impossible to navigate, so I wouldn't mind fixing it later.

> You can no longer use your mouse to resize rows and columns as you could before.

While you could in master, it wouldn't actually save anything, so I'd rather remove it and implement it properly separately.

File: packages/block-library/src/table/edit.js
Comment (iseulde) [reply packages/block-library/src/table/edit.js]: Why -1? Inserting one after should be +1 and right before 0?

File: packages/block-library/src/table/edit.js
Comment (talldan) [reply packages/block-library/src/table/edit.js]: React's warnings and errors mention the component name (based on the class or function name), so it can be really helpful.

File: packages/block-library/src/table/edit.js
Comment (iseulde) [reply packages/block-library/src/table/edit.js]: You'd still have to create the function though to pass the cell location.

File: packages/block-library/src/table/edit.js
Comment (iseulde) [reply packages/block-library/src/table/edit.js]: Since it contains a `RichText` that needs to update attributes, I think it make sense to have a context. Otherwise we can pass `attributes` and `setAttributes` as wel. I don't mind either way...

File: packages/blocks/src/api/parser.js
Comment (iseulde) [reply packages/blocks/src/api/parser.js]: There's no tests for the other sources either. Could add tests for all of them though... Not sure if valuable.

File: packages/block-library/src/table/edit.js
Comment (iseulde) [reply packages/block-library/src/table/edit.js]: I've left this as it is currently in master. Does it need a change? There's nothing that can be resized atm either.

File: packages/block-library/src/table/edit.js
Comment (iseulde) [reply packages/block-library/src/table/edit.js]: Getting the following error when adding `autoFocus`:

```
 error  The autoFocus prop should not be used, as it can reduce usability and accessibility for users
```

Should this be something that is part of writing flow instead, that is, always focussing content when a block is created?

**Comment by iseulde** [comment]
Make some adjustments and comments. Would it be possible to have another look at this?

File: packages/block-library/src/table/edit.js
Comment (noisysocks) [inline packages/block-library/src/table/edit.js]: `rowIndex` is bound to `createOnInsertRow`'s lexical scope, but incremented when the returned anonymous function is called. This might result in a bug where clicking _Insert Row_ inserts a new row two down from the current row because `rowIndex` was incremented twice.

I think that, in practice, this bug won't happen because inserting a row will cause React to call `render()` which in turn calls `createOnInsertRow()` and re-initialises `rowIndex`. 

Still, I think it's safer to not rely on this behaviour and instead avoid the mutating `+=` operator here.

```js
const { attributes, setAttributes } = this.props;
setAttributes( insertRow( attributes, { section, rowIndex: rowIndex + delta } ) );
```

File: packages/block-library/src/table/edit.js
Comment (noisysocks) [reply packages/block-library/src/table/edit.js]: Yes, you're right, we should pass `0` here. My mistake! 🙂

File: packages/block-library/src/table/edit.js
Comment (noisysocks) [reply packages/block-library/src/table/edit.js]: I didn't realise that we weren't saving the widths and heights in `master`. All good then!

File: packages/block-library/src/table/edit.js
Comment (noisysocks) [reply packages/block-library/src/table/edit.js]: Oh. Sure, sounds good 🙂

File: packages/blocks/src/api/parser.js
Comment (noisysocks) [reply packages/blocks/src/api/parser.js]: My gut says we should have them since this is an API we expose to third party block authors. But if there's none in `master` then we can do this separately! 👌

File: packages/block-library/src/table/index.js
Comment (noisysocks) [reply packages/block-library/src/table/index.js]: Works for me! 👍

File: packages/block-library/src/table/edit.js
Comment (talldan) [reply packages/block-library/src/table/edit.js]: Something like this could be a trade-off which solves both the binding and misdirection issues. It means binding more methods in the constructor, but I think it's easier to read:
```
insertRow( delta ) {
    if ( ! this.state.selectedCell ) {
        return;
    }

    const { section, rowIndex } = this.state.selectedCell;
    const { attributes, setAttributes } = this.props;
    const newRowIndex = rowIndex + delta;

    setAttributes( insertRow( attributes, { section, rowIndex: newRowIndex } ) );
    this.setState( { selectedCell: null } );
}

onInsertRowBefore() {
     this.insertRow(0);
}

onInsertRowAfter() {
     this.insertRow(1);
}

getTableControls() {
    return [
        {
	    icon: 'table-row-before',
	    title: __( 'Add Row Before' ),
	    isDisabled: ! selectedCell,
	    onClick: this.onInsertRowBefore,
        },
        //...
    ];
}
```

Anyway, that's just a stylistic thing. I think the bigger issue is the use of a closure to access `selectedCell`, which can be avoided given this.state.selectedCell can be accessed directly in those functions. That could be the cause of the bug mentioned by @noisysocks.

File: packages/block-library/src/table/edit.js
Comment (talldan) [reply packages/block-library/src/table/edit.js]: Cheers - I didn't know that.

File: packages/block-library/src/table/edit.js
Comment (gziolo) [reply packages/block-library/src/table/edit.js]: plus 💯 to what @talldan shared. In particular, having methods like `onInsertRowBefore` and `onInsertRowAfter` improves readability of the code by replacing indexes with the description of their functionality.

File: packages/block-library/src/table/edit.js
Comment (gziolo) [reply packages/block-library/src/table/edit.js]: I also think that it would be a good idea to extract this method to its own file and promote it to an independent component. It will decrease the file size of the current `edit` implementation and will help to better understand which props are used by those `Section` components.

File: packages/block-library/src/table/editor.scss
Comment (gziolo) [inline packages/block-library/src/table/editor.scss]: Nice to see those TinyMCE specific rules go away 👍

File: packages/block-library/src/table/index.js
Comment (gziolo) [inline packages/block-library/src/table/index.js]: I don't think we import SCSS files this way anymore. I bet this is merge issue.

File: packages/block-library/src/table/index.js
Comment (gziolo) [reply packages/block-library/src/table/index.js]: This is cool given that is backward compatible 👍

File: packages/block-library/src/table/test/__snapshots__/index.js.snap
Comment (gziolo) [inline packages/block-library/src/table/test/__snapshots__/index.js.snap]: The name of the test is probably wrong 😅

File: packages/block-library/src/table/test/__snapshots__/index.js.snap
Comment (gziolo) [inline packages/block-library/src/table/test/__snapshots__/index.js.snap]: This is much nicer than providing the default 2x2 table 💯

**Comment by gziolo** [comment]
screen shot 2018-08-09 at 13 50 09

@jasmussen - can you confirm it is good to go in that form?

**Comment by gziolo** [comment]
@iseulde this PR looks pretty close to be ready. Can you fix all the small issues remaining and open follow-up issues related to refactorings like:
- `renderSection` methods to component
- `onInsertRowBefore` and `onInsertRowAfter` methods

I would love to see those changes but they don't impact the UI at all. I can even help with those refactoring to speed up things as I want us to move forward with #7890 which I believe this PR blocks at the moment.

**Comment by iseulde** [comment]
Yes, I'll update this ASAP so we can move on with it.

**Comment by iseulde** [comment]
@jasmussen After your commit, there is an error: Warning: validateDOMNesting(...): <div> cannot appear as a descendant of <p>. Should we use something else?

**Comment by iseulde** [comment]
@jasmussen Yeah, maybe `div` elements don't really belong it a component that's just a wrapper for an input element.

**Comment by iseulde** [comment]
I reverted the commit for now

**Comment by jasmussen** [comment]
That's okay, but we do need to add back the font size and font family to this part before we can merge:

screen shot 2018-08-21 at 09 05 46

I'd rather ship with the previous change, but just remove the paragraphs.

**Comment by iseulde** [comment]
@jasmussen Would you be up for adding back what you think is needed?

**Comment by iseulde** [comment]
@noisysocks @talldan @gziolo Is there anything else I can address in this PR?

**Comment by gziolo** [comment]
i'll give it a spin tomorrow and do the final code check.

**Comment by aduth** [comment]
image

I don't see a single line of inline documentation in this +1047 changeset?

**Comment by talldan** [comment]
Thanks for the changes 😄

Looks good if those styling issues can be resolved. The point about documentation might be good to address as well.

I'll dismiss my review so that the others who have mentioned things can approve when they're happy.

**Comment by iseulde** [comment]
> I don't see a single line of inline documentation in this +1047 changeset?

Yes, I should add proper documentation, regardless of how big a changeset is, sorry about that.

**Comment by jasmussen** [comment]
Pushed polish again:

screen shot 2018-08-22 at 12 33 11

**Comment by iseulde** [comment]
Updated with docs and rebased.

**Review by noisysocks — APPROVED** [review]
Looks great, and works well in my testing 👍:shipit:

File: packages/block-library/src/table/edit.js
Comment (gziolo) [inline packages/block-library/src/table/edit.js]: I don't think this is a browser feature, it is part of the language:

screen shot 2018-08-23 at 08 14 10

File: packages/block-library/src/table/edit.js
Comment (gziolo) [inline packages/block-library/src/table/edit.js]: This should be `{number}` type :)

File: packages/block-library/src/table/index.js
Comment (gziolo) [inline packages/block-library/src/table/index.js]: I don't think we add spaces after dependencies section comment in other places. I don't think we have a rule for that, but sharing as an observation. Personally, I would let `prettier` to take care of formatting 😄

File: packages/block-library/src/table/test/state.js
Comment (gziolo) [inline packages/block-library/src/table/test/state.js]: Can we wrap `table` and `tableWithContent` with `deepFreeze` call to ensure there is no mutation?

**Comment by gziolo** [comment]
@iseulde - I addressed my own feedback.

There is one thing I noticed. When no cell is selected, I can see all actions in the dropdown menu, they are active but don't work - which is expected, however, we should update UI to reflect it.

**Comment by gziolo** [comment]
I did some debugging. `<DropdownMenu controls={ controls } />` - it turns out that each control takes only 3 props and `isDisabled` isn't one of them.

**Comment by gziolo** [comment]
I think I addressed my previous comment 19b7cb3. It allows to greatly simplify event callbacks, I'm not sure if you like this or not. We can revert part of the changes if you feel like double safety is expected.

**Comment by gziolo** [comment]
I added one more commit b9c92ed. It fixes the following use case:
- create a new table
- remove column
- remove column

Before the patch, they block view was empty.
After the patch, we render the view to create a table again.

**Review by gziolo — APPROVED** [review]
This is looking great! I really like the new UX introduced with this PR. Awesome work @iseulde 👏 

I applied a few commits, so feel free to review, update, rework them if you don't agree with all my changes.

File: packages/block-library/src/table/edit.js
Comment (iseulde) [inline packages/block-library/src/table/edit.js]: Hm, but `selectedCell` can still be `null`?

File: packages/block-library/src/table/edit.js
Comment (iseulde) [inline packages/block-library/src/table/edit.js]: I added these lines because that was @aduth's preference some time ago. I don't really mind, but it is not a doc block... So maybe it's better with the line, or just one asterisk at the start.

File: packages/block-library/src/table/edit.js
Comment (gziolo) [reply packages/block-library/src/table/edit.js]: We check it when providing `isDisabled` for menu items, then they are disabled. In addition, I added return clause in the even callback.

File: packages/block-library/src/table/edit.js
Comment (gziolo) [reply packages/block-library/src/table/edit.js]: If we want to change it, let's update in one batch all of them to match one style.

**Comment by gziolo** [comment]
ca4af7c - yeah, looks good 👍

**Comment by mcsf** [comment]
I like playing with this block a lot. One thing I'm noticing is the lack of alignment between `thead` and `tbody` if fixed width is off:

gutenberg-table-sections-alignment

Should I open an issue for this?

**Comment by tofumatt** [comment]
Looks like a bug to me!

**Comment by designsimply** [comment]
I filed an issue for the `thead` misalignment issue ([ref](https://github.com/WordPress/gutenberg/pull/8767#issuecomment-417854831)) at https://github.com/WordPress/gutenberg/issues/9779.

**Pull request merged**
